### ISSUE TITLE
add pluralsight.com extractor

### ISSUE THREAD

**Issue by chovy** [rg3/youtube-dl#6090]
I'd like to download some videos from pluralsight.com

**Comment by jaimeMF** [comment]
[Post an example url](https://github.com/rg3/youtube-dl/blob/533f67d3fa76eb44799276656207a25ece6d20a3/README.md#bugs)

**Comment by chovy** [comment]
You would need credentials to download from pluralsight.

**Comment by jaimeMF** [comment]
It doesn't matter, please post an example url so that if someone wants to work on it they can easily have an example to start with.

**Comment by anthonyettinger** [comment]
Video player window:
http://www.pluralsight.com/training/player?author=jeremy-foster&name=nodejs-internet-of-things-using-intel-edison-m1&mode=live&clip=0&course=nodejs-internet-of-things-intel-edison

Course page:
http://www.pluralsight.com/courses/nodejs-internet-of-things-intel-edison

**Comment by edwardchanjw** [comment]
Pluralsight have weekly free course, that a good start for non-credential code testing:
https://get.pluralsight.com/free-weekly-course.html

**Comment by edwardchanjw** [comment]
I don't know is this still work, a free subset subscription for PluralSight

https://offers.pluralsight.com/technet?loc=zTS3z&prod=zOTprodz&tech=zOttechz&prog=zOTprogz&type=zSOz&media=zOTmediaz&country=zUSz

**Comment by dstftw** [comment]
Pluralsight will be supported in the next version.

**Comment by chovy** [comment]
awesome.

**Comment by poudigne** [comment]
is the pluralsight downloader released yet ?

**Comment by chovy** [comment]
Here is an example course page:
http://www.digitaltutors.com/tutorial/528-Beginners-Guide-to-Digital-Painting-in-Photoshop

Here is an example video page:
http://www.digitaltutors.com/tutorial/528-Beginners-Guide-to-Digital-Painting-in-Photoshop#play-9891

**Comment by kasibkismath** [comment]
youtube-dl downloads for Pluralsight is not working from today

**Comment by chovy** [comment]
+1 for a fix

**Comment by tfrangio** [comment]
First Thanks for the ***youtube-dl*** it ***ROCKS!***

Can some one post the correct command to download from Pluralsight once it is fixxed. Right now I am using:
```
youtube-dl --username <myusername> --password <mypassword> <Pluralsight Video URL>
```
I am on windows 7 - 64 bit. Here is the result of the ***youtube-dl.exe -v***:
```
C:\Users\efrangio\Downloads>youtube-dl.exe -v
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'-v']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2016.01.15
[debug] Python version 2.7.10 - Windows-7-6.1.7601-SP1
[debug] exe versions: none
[debug] Proxy map: {}
Usage: youtube-dl.exe [OPTIONS] URL [URL...]

youtube-dl.exe: error: You must provide at least one URL.
Type youtube-dl --help to see a list of all options.

C:\Users\efrangio\Downloads>
```


I get this error:
```
[generic] player?course=linux-managing-dns-servers-lpic-2: Requesting header
WARNING: Could not send HEAD request to https://app.pluralsight.com/player?cours
e=linux-managing-dns-servers-lpic-2: HTTP Error 400: Bad Request
[generic] player?course=linux-managing-dns-servers-lpic-2: Downloading webpage
ERROR: Unable to download webpage: HTTP Error 400: Bad Request (caused by HTTPEr
ror()); please report this issue on https://yt-dl.org/bug . Make sure you are us
ing the latest version; type  youtube-dl -U  to update. Be sure to call youtube-
dl with the --verbose flag and include its complete output.
'author' is not recognized as an internal or external command,
operable program or batch file.
'name' is not recognized as an internal or external command,
operable program or batch file.
ERROR: Invalid argument/option - '=0'.
Type "CLIP /?" for usage.
Invalid parameter - =live
```
Any ideas on how to get this working would be infinitely appreciated?

**Comment by yan12125** [comment]
Looks like you need some more quotation marks in your command. See [this FAQ entry](https://github.com/rg3/youtube-dl/#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command). If the problem is still, feel free to open a new issue with the full command you're using and the output.

**Comment by tfrangio** [comment]
Thanks for your quick help. I tried to put the url in double quotes like this:
```
youtube-dl --username <myusername> --password <mypassword> "<Pluralsight Video URL>"
```
Here was the result:
```
[generic] player?course=linux-managing-dns-servers-lpic-2&author=andrew-mallett&
name=linux-managing-dns-servers-lpic-2-m1&clip=0&mode=live: Requesting header
WARNING: Falling back on generic information extractor.
[generic] player?course=linux-managing-dns-servers-lpic-2&author=andrew-mallett&
name=linux-managing-dns-servers-lpic-2-m1&clip=0&mode=live: Downloading webpage
[generic] player?course=linux-managing-dns-servers-lpic-2&author=andrew-mallett&
name=linux-managing-dns-servers-lpic-2-m1&clip=0&mode=live: Extracting informati
on
ERROR: Unsupported URL: https://app.pluralsight.com/player?course=linux-managing
-dns-servers-lpic-2&author=andrew-mallett&name=linux-managing-dns-servers-lpic-2
-m1&clip=0&mode=live
```
This was definitely one of my issues but I am getting a ***Unsupported URL*** error now.

**Comment by mousepotato** [comment]
You may not be able to download using a specific video URL page, but the course URL page is OK.
e.g.,  youtube-dl --username <myusername> --password <mypassword> "https://app.pluralsight.com/library/courses/using-wmi-and-cim-powershell"

**Comment by tfrangio** [comment]
Thanks a bunch mousepotato,

The solution to my issue was to use the ***course URL page*** in double quotes just as shown above.  As soon as I did that I got the videos to download.

**Comment by vnpnlz** [comment]
This doesn't work for me.

**Comment by vnpnlz** [comment]
This doesn't work for me. I got an error Unable to download the page : <urlopen error [Errno 8]]> __ssl.c:510 EOF occured in violation of protocol>

**Comment by trashbat** [comment]
Be warned that you may find yourself locked out of your account if you download too much data from them, this is the email I received earlier:

> We have blocked your account because our security systems have flagged your Pluralsight account for an unusual amount activity. This does mean a high volume of requests that are in the realm of a request every 10-30 seconds for a prolonged period of time. Please note that this high volume of activity is in violation of our terms of service [https://www.pluralsight.com/terms].

In my case I'd just downloaded a single course and I was halfway through downloading a second one. Easy enough to get it unblocked, though, you just have to email support.

**Comment by dstftw** [comment]
@trashbat what version have you used? Post the output of `youtube-dl -v`.

**Comment by trashbat** [comment]
@dstftw I was using latest master, output is `2016.01.27`. It did pause between each video to avoid throttling, but Pluralsight still didn't like the volume of data that I was downloading from them (~1.1GB from 2 courses when they locked my account).

**Comment by dstftw** [comment]
Have you downloaded courses in parallel? What was the `--sleep-interval` you used? There is already a [anti-throttling sleep](https://github.com/rg3/youtube-dl/blob/master/youtube_dl/extractor/pluralsight.py#L214) that previously worked in most of the cases.

**Comment by trashbat** [comment]
No I downloaded them sequentially, and I didn't specify a `--sleep-interval`. Perhaps I was just unlucky; the first course that I downloaded was rather large (691MB), and I managed to get 426M of the second one downloaded before my account was locked. Could also be due to my fast connection speed at work (~750 Mbit/s), which meant that the data was downloaded in a short space of time.

**Comment by stevegreen22** [comment]
Hey all, I've tried the suggested solution to the above and even used the example that was posted but I'm still getting an error.  Any suggestions on what I could be doing wrong?

~ > youtube-dl --username myusername -p my password "https://app.pluralsight.com/library/courses/using-wmi-and-cim-powershell"

[generic] using-wmi-and-cim-powershell: Requesting header
[redirect] Following redirect to https://app.pluralsight.com/id/signin/?redirectTo=https%3A%2F%2Fapp.pluralsight.com%2Flibrary%2Fcourses%2Fusing-wmi-and-cim-powershell
[generic] ?redirectTo=https://app.pluralsight: Requesting header
WARNING: Falling back on generic information extractor.
[generic] ?redirectTo=https://app.pluralsight: Downloading webpage
[generic] ?redirectTo=https://app.pluralsight: Extracting information
ERROR: Unsupported URL: https://app.pluralsight.com/id/signin/?redirectTo=https%3A%2F%2Fapp.pluralsight.com%2Flibrary%2Fcourses%2Fusing-wmi-and-cim-powershell

**Comment by tfrangio** [comment]
I had a problem with getting locked out of this site as well. I found that crawling the site like a snail has worked for me. As I am in no rush to download an entire course this is fine with me. I set this in the morning and by the end of the day I have a course downloaded. I did this because I had already been locked out once and did not want to have that happen again.

The two parameters that dictate the speed are "--rate-limit 50K" and "--sleep-interval 600" notice that I am only fetching another video every 10 min with the sleep-interval set to 600.

    youtube-dl.exe --verbose  --username <<YOUREMAIL>> --password <<YOURPASSWORD>> --rate-limit 50K --sleep-interval 600 -o "%(autonumber)s - %(title)s.%(ext)s" "https://app.pluralsight.com/library/courses/bash-shell-scripting"

**Comment by bruinebeer** [comment]
@tfrangio hi, i am using your settings and it works, but it is really really slow.. the 600 sleep-interval seems excessive though? would it hurt to make it wait less? and maybe speed up a lil bit?

**Comment by tfrangio** [comment]
@bruinebeer I guess you could experiment with the values of "sleep-interval" and "rate-limit" but you will risk having your account shut off. Sleep - interval is specified on seconds so 600 is 10 minutes you could try 300 that would be every 5 min.

**Comment by bruinebeer** [comment]
@tfrangio Thx! Is it possible to continue a download when it gets interrupted?

**Comment by chovy** [comment]
I'm curious about that too. If there is a `--resume` flag where it would start from the last incomplete file.

**Comment by N1t1nA** [comment]
@mousepotato Thanks!

**Comment by xikaos** [comment]
I'm trying to download a course only using the sleep interval flag, cause to me it seems like there is something to do with request frequency/number.  Obviously you will spent like 5 to 10 minutes per video, so requesting their URL so often will make their security alert scream and you will probably be blocked. I have a trial account so as soon I finnish downloading this one I'll move to the next and will keep you guys tuned about my progress.
As soon as I finnish downloading this course, I'll move to the next and let you guys know what happened, cause I have a trial account. Nothing to lose.

**Comment by chovy** [comment]
i found i could download 2-3 courses, then I got banned. So be careful.

**Comment by xikaos** [comment]
@chovy Did you used sleep interval? I wonder how the hell they can make the assumption that you are a script, not a regular person watching the videos for like 5 minutes then going to the next one.

**Comment by chovy** [comment]
I think so, it waited 10 minutes between courses. I used whatever the default behavior is.

**Comment by bruinebeer** [comment]
@chovy With tfrangio's options i don't think pluralsight will notice it. It's takes almost a full day to download a course.

**Comment by chovy** [comment]
is there anyway to resume? i tried doing this but my laptop went to sleep on me

**Comment by xikaos** [comment]
@chovy DId you tried the -c option?  I think if you dont use the overwrite option, maybe it continues from where it already downloaded.

**Comment by chovy** [comment]
I did not pass any options like `-c` -- is that required? or is it smart enough to just resume from the last incomplete file?

**Comment by xikaos** [comment]
When you pass the -c option it will skip the already downloaded files and contine from the last one you downloaded. I'm not sure about how the script treats this last file you downloaded if you don't supply the --no-overwrites option.

**Comment by xikaos** [comment]
I'm downloading my sixth course and was not banned yet.
It is safe to download at full speed, but with a --sleep-interval of 5 minutes. I really think that this interval can be of 2 minutes, but I'm afraid to test while I still have 27 days of subscription.
Will try on my last 3 days and keep you guys informed of what happened.

**Comment by N1t1nA** [comment]
hi @xikaos can you please provide the syntax of (safe to download at full speed --sleep-interval of 5 minutes)

**Comment by xikaos** [comment]
<code>youtube-dl --verbose --username username --password password
--sleep-interval 300 "https://app.pluralsight.com/courses/css3-in-depth
"</code>

**Comment by maestrofx** [comment]
hi @xikaos, which version do you use to work well?

**Comment by xikaos** [comment]
@maestrofx [debug] youtube-dl version 2016.02.13. 15GB of videos downloaded so far, no bans.

**Comment by maestrofx** [comment]
Ok, I am in a new VPS to try this with latest version (2016.02.22). But, I have a weird error [1]. 
I assumed this latest version not working, so I have downgrade to 2016.02.13. But, the same error come.

Is this issue for youtube-dl or anything else?

Thanks




[1] error

> ERROR: Unable to download JSON metadata: HTTP Error 403: Forbidden (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 354, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 1905, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python2.7/urllib2.py", line 410, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 523, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 448, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 531, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)

**Comment by xikaos** [comment]
@maestrofx I actually didn't knew I was using an outdated version of youtube-dl, but your error seems like authentication problem.
 It will sound incredibly stupid of my part, but double-check your credentials and the options you pass to the script before executing. 
If that does not solve the problem, open an issue here on Github.

**Comment by maestrofx** [comment]
Ok, here are my updates:

- I've been still playing with some versions to check which one is working for me
- I've read some issues and comments related to this error and come back later to update
- no problem with the authentication problem or credentials or another params to the script
- I can open videos with the player, so no problem with banned status by pluralsight.
- Also, I've tried this with one of my banned account.

Thanks

**Comment by mirfilip** [comment]
@xikaos 
>  I wonder how the hell they can make the assumption that you are a script, not a regular person watching the videos for like 5 minutes then going to the next one.

Looking at the explaination they give you when your account gets banned, it's a matter of
> high volume of requests that are in the realm of a request every 10-30 seconds for a prolonged period of time

which
> this high volume of activity is in violation of our terms of service

So, I would recommend to set a
- good User Agent
- `sleep-interval` to something like ~120 sec

**Comment by xikaos** [comment]
@mirfilip Thanks for the rich response, sir! I'm actually very satisfied with youtube-dl functionalities. It would be nice if the script could track precisely the status of the playlist instead of making the requests one-by-one and then skipping the files you already have downloaded. Its a shame I don't know Python to retrieve the favor of the tool giving me 30GB worth of videos for offline access! Thanks to all who contributed to this wonderful tool.

**Comment by readergirltracey** [comment]
Anybody tried with 120s?

**Comment by maestrofx** [comment]
@readergirltracey try `--sleep-interval` 300, it works well. But, your account will be absolutely blocked if you run the command over and over again due to a file that's skipped in the playlist during the process.

Say the playlist has 32 files and the 10th file was skipped. When you run the command again, it's risky to block by Pluralsight's system.

**Comment by damianopetrungaro** [comment]
It would be awesome if also the "skills" url could be grabbed

Link sample https://app.pluralsight.com/paths/skill/docker

Of course it need auth =)

**Comment by yan12125** [comment]
@damianopetrungaro Please open a new issue for that.

**Comment by Ebram-Tharwat** [comment]
Can anyone please, share his experience when setting `--sleep-interval`  set to **120s** ?
Is **120s** enough or it should be more longer?

**Comment by damianopetrungaro** [comment]
120 is ok.

**Comment by Ebram-Tharwat** [comment]
@damianopetrungaro , do you use `--rate-limit` option to limit your download speed or you use your max speed (if max speed, can you mention it)?
Also, do you download course**s** sequentially or in parallel? 

Thank you in advance.

**Comment by damianopetrungaro** [comment]
@Ebram-Tharwat  i used `--sleep-interval` only at 120 sequentially.

**Comment by Anil86** [comment]
Past 1 day, downloading interrupts after every 10 - 15 mins.
Sleep Interval  =  120 (same issue with 180)
Download speed  =  2 Mbps

Anyone else facing same issue?

**Comment by sonhoang0611** [comment]
120s works fine for me tks. I used: youtube-dl --username yourname  --password yourpassword --verbose --sleep-interval 120s --all-subs "yourlink"

**Comment by shekhtman** [comment]
Recently I was redirected, after login, to a page asking me to accept/decline their terms of use. Did any one get this?
![pluralsight](https://cloud.githubusercontent.com/assets/8742187/23828056/98e6b40e-0679-11e7-9cc5-f91c15ddcc27.png)

**Comment by yan12125** [comment]
@shekhtman There are quite a few similar reports. You'll need to accept the ToS manually in browsers.

**Comment by mutuadavid93** [comment]
Just use BigaSoft Video Downloader and download not more than 6hrs videos length a day. You will be fine.
